### ISSUE TITLE
[QUERY] Message size doubled after using Spring Cloud Stream to send message to Azure Event Hub

### ISSUE THREAD

**Issue by w-lin** [Azure/azure-sdk-for-java#25076]
We have an existing system running in production to send messages to Azure event hub WITHOUT using spring-cloud-stream. Here is the code
```
EventData eventData = new EventData(myString);  //myString.length is about 999,000
EventDataBatch batch = eventHubProducerClient.createBatch();
boolean isAdded = batch.tryAdd(eventData);
....
eventHubProducerClient.send(batch)
```

Event hub supports maximum message size of 1MB, since the source code of EventData uses UTF-8 encoding thus each character in the string becomes 1 byte. So we assembled each message to be very close to 1,000,000 characters
```
package com.azure.messaging.eventhubs;
public class EventData {
    ...
    public EventData(String body) {
        this(((String)Objects.requireNonNull(body, "'body' cannot be null.")).getBytes(StandardCharsets.UTF_8));
    }
   ...
}
```

However when we are rewriting above code to use spring-could-stream, we get the exception of exceeding max size 1024 kb 
```
@Bean
public Function<Message<?>, Message<String>> produce {
return m-> { 
...
MessageBuilder.withPayload( myString).build(); //myString.length is 999,000 and we get exception
...
}
```
```
2021-10-27 21:04:42.092 ERROR 1 --- [z2-accessflow-2] c.a.s.i.e.impl.AbstractEventHubTemplate  : Event is larger than maximum allowed size. Exception: com.azure.core.amqp.exception.AmqpException: Size of the payload exceeded maximum message size: 1024 kb, errorContext[NAMESPACE: dev-qa-janus-events-cus.servicebus.windows.net, PATH: dev_qa_touchstone_accessflow_event, REFERENCE_ID: 
```
I tested using a string of length=540,000 and it failed too, but a string of length=490,000 went through successfully. 
So during rewriting, I have to cut down each message size to half of 999,000 characters to avoid the error above.

However when we run this new code in testing environment, the throughput and number of messages are doubled compared with production system, given that the same data are sent to both existing system (production) and new system (testing environment). Take real statistics data from a event hub from Azure dashboard graphs for example, given a one-hour window:
production-event-hub-a:  total throughput 20GB and total number of messages 20K (each message size = 20GB/20K=1MB)
testing-event-hub-a (use spring-clound-stream):  total throughput 40GB and total number of messages 40K (each message size = 40GB/40K=1MB).

To sum up, during rewriting to use spring-could-stream, we have to cut down each message (string length) to below 500,000  characters (half of what we had in existing system), but each message is still 1MB according to Azure statistics graph thus throughput is doubled. 

I am not sure if I miss something, I wonder if there is a way to send the messages using spring-clound-stream to match the same throughput of our existing system?

**Comment by jialigit** [comment]
@w-lin the thing is this, after running the sample(https://github.com/Azure-Samples/azure-spring-boot-samples/blob/3289e3498514bacd9e2e6f572412e069932f9992/eventhubs/azure-spring-cloud-stream-binder-eventhubs/eventhubs-binder/src/main/java/com/azure/spring/sample/eventhubs/binder/EventProducerConfiguration.java) with the message size to be   999,000 cannot reproduce the exception in my environment, so would you please show some details of the way you send the message?  and the following is the snippit i used to produce the message. Any suggestions are welcome.


```
@Bean
    public Supplier<Message<String>> supply() {
        return () -> {
            LOGGER.info("Sending message, sequence " + i);
            StringBuffer message = new StringBuffer();
            for(int i=0;i<999000;i++){
                message.append("1");
            }
     
            LOGGER.info("message size = {}", message.toString().length());
 
            return MessageBuilder.withPayload(message.toString()).build();
        };
    }

```

**Comment by w-lin** [comment]
Hi @jialigit  I copy your code ( for(int i=0;i<999000;i++) message.append("1"); returnMessageBuilder.withPayload(message.toString()).build();) and I get the exception

```
2021-11-01 19:14:14.714 ERROR 1 --- [z2-borderflow-1] c.a.s.i.e.impl.AbstractEventHubTemplate  : Event is larger than maximum allowed size. Exception: com.azure.core.amqp.exception.AmqpException: Size of the payload exceeded maximum message size: 1024 kb, errorContext[NAMESPACE: dev-qa-janus-events-cus.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: dev_qa_touchstone_borderflow_event, REFERENCE_ID: 4F17B9B2001EA0D5001EA0FC61803C11_G5S3, LINK_CREDIT: 300]
```

Here is my yaml file:
```
spring:
  cloud:
    stream:
      function:
        definition: borderflow_region2
      eventhub:
        bindings:
          borderflow_region2-out-0:
            producer:
              sync: true
      bindings:
        borderflow_region2-out-0:
          binder: borderflow-eventhub-2
          destination: {{ .Values.azure.borderflowTopic }}
      binders:
        borderflow-eventhub-2:
          type: eventhub
          default-candidate: false
```

I also upgrade to latest versions of azure and spring library 
```
		<dependency>
			<groupId>com.azure.spring</groupId>
			<artifactId>azure-spring-cloud-stream-binder-eventhubs</artifactId>
			<version>2.9.0</version>
		</dependency>

	<dependencyManagement>
		<dependencies>
			<dependency>
				<groupId>org.springframework.cloud</groupId>
				<artifactId>spring-cloud-dependencies</artifactId>
				<version>2020.0.4</version>
				<type>pom</type>
				<scope>import</scope>
			</dependency>
		</dependencies>
	</dependencyManagement>
```

I use JDK 11.

Are you able to spot any difference between my code and your testing code?

**Comment by jialigit** [comment]
@w-lin hi, I am also wondering what the function definiton (shown in above yaml)  'borderflow_region2' is ? Function or Supplier? It is seemingly the yaml is not complete and I cannot see the whole picture clearly. I have tried to reproduce the situation but cannot make that exception 'jump out'. I do missed something and get stucked for now, so should you give a complete demo if it is convenient? many thanks :)

**Comment by w-lin** [comment]
@jialigit The rest yaml file (and whether it is function or supplier) is irrelevant. Our yaml file is not small that is why I did not paste other lines that are irrelevant. But here is full code 

configmap.yaml
```
kind: ConfigMap
apiVersion: v1
metadata:
  name: plx-azpublisher
  namespace: {{ .Values.namespace }}
data:
  application.yaml: |-

    dcswitch:
      location: {{ .Values.global.location }}
      url: {{ .Values.dcswitch.url }}
    management:
      endpoint:
        health:
          probes:
            enabled: true
          livenessState:
            enabled: true
    server:
      port: 8082
    spring:
      rabbitmq:
        host: rabbit.proxy.internal.local
        port: 5672
        ssl:
          enabled: false
        virtual-host: /plx_mpc
      cloud:
        stream:
          function:
            definition: tophits_region1;tophits_region2;borderflow_region2;accessflow_region1;accessflow_region2;generic_region1;generic_region2;attack_region1;attack_region2;dcSwitch
          eventhub:
            bindings:
              borderflow_region1-out-0:
                producer:
                 sync: true
              borderflow_region2-out-0:
                producer:
                  sync: true
              accessflow_region1-out-0:
                producer:
                 sync: true
              accessflow_region2-out-0:
                producer:
                  sync: true
              tophits_region1-out-0:
               producer:
                 sync: true
              tophits_region2-out-0:
                producer:
                  sync: true
              attack_region1_out-0:
               producer:
                 sync: true
              attack_region2-out-0:
                producer:
                  sync: true
              generic_region1-out-0:
               producer:
                 sync: true
              generic_region2-out-0:
                producer:
                  sunc: true
          rabbit:
            default:
              republishToDlq: false
              ttl: 120000
            bindings:
              dcSwitch-in-0:
                consumer:
                  binding-routing-key: datacenter.switch
                  declare-exchange: false
                  anonymousGroupPrefix: dcswitch.cs.azpublisher.
          bindings:
            dcSwitch-in-0:
              destination: plx.metric.exchange.topic
            tophits_region1-in-0:
              destination: plx.azure.tophits
              group: janus-az1-tophits
              consumer:
                maxConcurrency: {{ .Values.janus.tophits.consumer.maxConcurrency }}
                concurrency: {{ .Values.janus.tophits.consumer.concurrency }}
            tophits_region1-out-0:
              binder: tophits-eventhub-1
              destination: {{ .Values.azure.tophitsTopic }}
            tophits_region2-in-0:
              destination: plx.azure.tophits
              group: janus-az2-tophits
              consumer:
                maxConcurrency: {{ .Values.janus.tophits.consumer.maxConcurrency }}
                concurrency: {{ .Values.janus.tophits.consumer.concurrency }}
            tophits_region2-out-0:
              binder: tophits-eventhub-2
              destination: {{ .Values.azure.tophitsTopic }}
            borderflow_region1-in-0:
              destination: plx.azure.border
              group: janus-az1-borderflow
              consumer:
                maxConcurrency: {{ .Values.janus.sflow.consumer.maxConcurrency }}
                concurrency:  {{ .Values.janus.sflow.consumer.concurrency }}
            borderflow_region1-out-0:
              binder: borderflow-eventhub-1
              destination: {{ .Values.azure.borderflowTopic }}
            borderflow_region2-in-0:
              destination: plx.azure.border
              group: janus-az2-borderflow
              consumer:
                maxConcurrency: {{ .Values.janus.sflow.consumer.maxConcurrency }}
                concurrency:  {{ .Values.janus.sflow.consumer.concurrency }}
            borderflow_region2-out-0:
              binder: borderflow-eventhub-2
              destination: {{ .Values.azure.borderflowTopic }}
            accessflow_region1-in-0:
              destination: plx.azure.access
              group: janus-az1-accessflow
              consumer:
                maxConcurrency: {{ .Values.janus.sflow.consumer.maxConcurrency }}
                concurrency: {{ .Values.janus.sflow.consumer.concurrency }}
            accessflow_region1-out-0:
              binder: accessflow-eventhub-1
              destination: {{ .Values.azure.accessflowTopic }}
            accessflow_region2-in-0:
              destination: plx.azure.access
              group: janus-az2-accessflow
              consumer:
                maxConcurrency: {{ .Values.janus.sflow.consumer.maxConcurrency }}
                concurrency: {{ .Values.janus.sflow.consumer.concurrency}}
            accessflow_region2-out-0:
              binder: accessflow-eventhub-2
              destination: {{ .Values.azure.accessflowTopic }}
            generic_region1-in-0:
              destination: plx.azure.generic
              group: janus-az1-generic
              consumer:
                maxConcurrency: {{ .Values.janus.generic.consumer.maxConncurrency }}
                concurrency: {{ .Values.janus.generic.consumer.concurrency }}
            generic_region1-out-0:
              binder: generic-eventhub-1
              destination: {{ .Values.azure.genericTopic }}
            generic_region2-in-0:
              destination: plx.azure.generic
              group: janus-az2-generic
              consumer:
                maxConcurrency: {{ .Values.janus.generic.consumer.maxConncurrency }}
                concurrency:  {{ .Values.janus.generic.consumer.concurrency }}
            generic_region2-out-0:
              binder: generic-eventhub-2
              destination: {{ .Values.azure.genericTopic }}
            attack_region1-in-0:
              destination: plx.azure.attack
              group: janus-az1-attack
              consumer:
                maxConcurrency: {{ .Values.janus.generic.consumer.maxConncurrency }}
                concurrency:  {{ .Values.janus.generic.consumer.concurrency }}
            attack_region1-out-0:
              binder: attack-eventhub-1
              destination: {{ .Values.azure.attackEventTopic }}
            attack_region2-in-0:
              destination: plx.azure.attack
              group: janus-az2-attack
              consumer:
                maxConcurrency: {{ .Values.janus.generic.consumer.maxConncurrency }}
                concurrency: {{ .Values.janus.generic.consumer.concurrency }}
            attack_region2-out-0:
              binder: attack-eventhub-2
              destination: {{ .Values.azure.attackEventTopic }}



          binders:
            rabbit-1:
              type: rabbit
              default-candidate: true
              environment:
                spring:
                  rabbitmq:
                    host: rabbit.proxy.internal.local
                    port: 5672
                    ssl:
                     enabled: false
                    virtual-host: /plx_mpc
            tophits-eventhub-1:
              type: eventhub
              default-candidate: false

            tophits-eventhub-2:
              type: eventhub
              default-candidate: false

            borderflow-eventhub-1:
              type: eventhub
              default-candidate: false

            borderflow-eventhub-2:
              type: eventhub
              default-candidate: false

            accessflow-eventhub-1:
              type: eventhub
              default-candidate: false

            accessflow-eventhub-2:
              type: eventhub
              default-candidate: false

            generic-eventhub-1:
              type: eventhub
              default-candidate: false

            generic-eventhub-2:
              type: eventhub
              default-candidate: false

            attack-eventhub-1:
              type: eventhub
              default-candidate: false

            attack-eventhub-2:
              type: eventhub
              default-candidate: false
```

deployment.yaml
```
apiVersion: apps/v1
kind: Deployment
metadata:
  namespace: {{ .Values.namespace}}
  name: {{ include "plx-azpublisher.fullname" . }}
  labels:
    {{- include "plx-azpublisher.labels" . | nindent 4 }}
spec:
{{- if not .Values.autoscaling.enabled }}
  replicas: {{ .Values.replicaCount }}
{{- end }}
  selector:
    matchLabels:
      {{- include "plx-azpublisher.selectorLabels" . | nindent 6 }}
  template:
    metadata:
      annotations:
        sidecar.istio.io/rewriteAppHTTPProbe: "true"
        proxy.istio.io/config: '{ "holdApplicationUntilProxyStarts": true }'
        sidecar.istio.io/userVolumeMount: '[{"name":"spock-certs", "mountPath":"/etc/spockcerts", "readonly":true},{"name":"spock-ca-certs", "mountPath":"/etc/spock-ca-certs", "readonly":true}]'
        sidecar.istio.io/userVolume: '[{"name":"spock-certs", "secret":{"secretName":"spock-tls-spock"}},{"name":"spock-ca-certs", "secret":{"secretName":"spock-ca-kdc-akamai"}}]'
        traffic.sidecar.istio.io/excludeOutboundPorts: "2181"
        sidecar.istio.io/logLevel: 'info'
      labels:
        {{- include "plx-azpublisher.selectorLabels" . | nindent 8 }}
    spec:
      {{- with .Values.imagePullSecrets }}
      imagePullSecrets:
        {{- toYaml . | nindent 8 }}
      {{- end }}
      serviceAccountName: {{ include "plx-azpublisher.serviceAccountName" . }}
      securityContext:
        {{- toYaml .Values.podSecurityContext | nindent 8 }}
      hostAliases:  # We need to fix dns
        - ip: "10.10.100.1"
          hostnames:
            - "rabbit.proxy.internal.local"
        - ip: "10.10.100.3"
          hostnames:
            - "webservice.proxy.internal.local"
      containers:
        - name: {{ .Chart.Name }}
          env:
            {{- range .Values.environmentVariables }}
            - name: {{ .name }}
              value: {{ .value }}
            {{- end }}
            - name: spring.cloud.stream.binders.tophits-eventhub-1.environment.spring.cloud.azure.eventhub.connection-string
              valueFrom:
                secretKeyRef:
                  name: azure-secret
                  key: tophitsConnectionStringR1
            - name: spring.cloud.stream.binders.tophits-eventhub-2.environment.spring.cloud.azure.eventhub.connection-string
              valueFrom:
                secretKeyRef:
                  name: azure-secret
                  key: tophitsConnectionStringR2
            - name: spring.cloud.stream.binders.borderflow-eventhub-1.environment.spring.cloud.azure.eventhub.connection-string
              valueFrom:
                secretKeyRef:
                  name: azure-secret
                  key: borderConnectionStringR1
            - name: spring.cloud.stream.binders.borderflow-eventhub-2.environment.spring.cloud.azure.eventhub.connection-string
              valueFrom:
                secretKeyRef:
                  name: azure-secret
                  key: borderConnectionStringR2
            - name: spring.cloud.stream.binders.accessflow-eventhub-1.environment.spring.cloud.azure.eventhub.connection-string
              valueFrom:
                secretKeyRef:
                  name: azure-secret
                  key: accessConnectionStringR1
            - name: spring.cloud.stream.binders.accessflow-eventhub-2.environment.spring.cloud.azure.eventhub.connection-string
              valueFrom:
                secretKeyRef:
                  name: azure-secret
                  key: accessConnectionStringR2
            - name: spring.cloud.stream.binders.generic-eventhub-1.environment.spring.cloud.azure.eventhub.connection-string
              valueFrom:
                secretKeyRef:
                  name: azure-secret
                  key: genericConnectionStringR1
            - name: spring.cloud.stream.binders.generic-eventhub-2.environment.spring.cloud.azure.eventhub.connection-string
              valueFrom:
                secretKeyRef:
                  name: azure-secret
                  key: genericConnectionStringR2
            - name: spring.cloud.stream.binders.attack-eventhub-1.environment.spring.cloud.azure.eventhub.connection-string
              valueFrom:
                secretKeyRef:
                  name: azure-secret
                  key: attackConnectionStringR1
            - name: spring.cloud.stream.binders.attack-eventhub-2.environment.spring.cloud.azure.eventhub.connection-string
              valueFrom:
                secretKeyRef:
                  name: azure-secret
                  key: attackConnectionStringR2


          securityContext:
            {{- toYaml .Values.securityContext | nindent 12 }}
          {{- if .Values.localDevelopment }}
          image: "{{ .Values.localImage }}"
          {{- else }}
          image: "{{ .Values.global.imageRegistry }}/{{ include "service.name" . }}/{{ include "service.name" . }}:{{ .Values.global.imageBuildTag }}"
          {{- end }}
          ports:
            - name: http
              containerPort: {{ .Values.service.targetPort }}
              protocol: TCP
          livenessProbe:
            httpGet:
              path: {{.Values.service.probes.livenessProbe.path}}
              port: {{.Values.service.port}}
            initialDelaySeconds: {{.Values.service.probes.livenessProbe.initialDelaySeconds}}
            periodSeconds: {{.Values.service.probes.livenessProbe.periodSeconds}}
          readinessProbe:
            httpGet:
               path: {{.Values.service.probes.readinessProbe.path}}
               port: {{.Values.service.port}}
            initialDelaySeconds: {{.Values.service.probes.readinessProbe.initialDelaySeconds}}
            periodSeconds: {{ .Values.service.probes.readinessProbe.periodSeconds}}
          startupProbe:
            httpGet:
              path: {{.Values.service.probes.readinessProbe.path}}
              port: {{ .Values.service.probes.startupProbe.port}}
            initialDelaySeconds: {{ .Values.service.probes.startupProbe.initialDelaySeconds}}
            failureThreshold: {{.Values.service.probes.startupProbe.failureThreshold}}
            periodSeconds: {{ .Values.service.probes.startupProbe.periodSeconds}}
{{/*          resources:*/}}
{{/*            requests:*/}}
{{/*              memory: {{ .Values.resources.requests.memory }}*/}}
{{/*              cpu: {{ .Values.resources.requests.cpu }}*/}}
{{/*            limits:*/}}
{{/*              memory: {{ .Values.resources.requests.memory }}*/}}
{{/*              cpu: {{ .Values.resources.requests.cpu }}*/}}
          command: [ '/{{ include "plx-azpublisher.fullname" . }}/entrypoint.sh','/{{ include "plx-azpublisher.fullname" . }}/{{ include "plx-azpublisher.fullname" . }}.jar' ]
      {{- with .Values.nodeSelector }}
      nodeSelector:
        {{- toYaml . | nindent 8 }}
      {{- end }}
      {{- with .Values.affinity }}
      affinity:
        {{- toYaml . | nindent 8 }}
      {{- end }}
      {{- with .Values.tolerations }}
      tolerations:
        {{- toYaml . | nindent 8 }}
      {{- end }}
```

```
    @Bean
    public Function<Message<?>, Message<String>> borderflow_region2() {
        return (m) ->  {
            String x= (String) m.getPayload();
            
            StringBuffer message = new StringBuffer();
            for(int i=0;i<999000;i++){
               message.append("1");
            }
            return MessageBuilder.withPayload(message.toString()).build();
        };
    }
```

**Comment by jialigit** [comment]
@w-lin hi, I have tried setup the character-set to be utf-16 in the code which can reproduce the exception as you mentioned in this issue. And I am doubting the real reason of the occurrence of the exception is really the 'CHARACTER SET'.

SO can you try setup the character-set as utf-8 explicitly in the code and see the result? the default is dependent on the platform you are using, so please check and try and tell us your testing result. thanks very much. :)

this is the snippet i used
```
message.append(new String("1".getBytes(), StandardCharsets.UTF_8));
```

**Comment by w-lin** [comment]
@jialigit  I use the code you provided above and still get the same error - Exception: com.azure.core.amqp.exception.AmqpException: Size of the payload exceeded maximum message size: 1024 kb,

**Comment by jialigit** [comment]
@w-lin,  please check the following in your testing env:

1. check the default encoding in your testing env
```    
Charset cs = Charset.defaultCharset();
System.out.println(cs);
```
please reassure the default platform charset, I think this is point...

2. And try the following code :
```

for(int i=0;i<540000;i++){
                //set the charset UTF-8 at both places
                message.append(new String("1".getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8));                
 }
```

3.  print the value of myStr.getBytes(StandardCharsets.UTF_8).length to reassure the size.
 

I tested explicitly set the charset to be UTF-16 in the first place, and get the exception, and change it to UTF-8, then goes ok.  
 ```
message.append(new String("1".getBytes(StandardCharsets.UTF_16), StandardCharsets.UTF_8));
 ```

**Comment by w-lin** [comment]
@jialigit I tried several different scenarios, there are the outputs


Not sure where the "Cannot send an EventBatch that is empty" is from (and why the log shows EventHubProducerAsyncClient whereas our yaml file set  borderflow_region2-out-0.producer.sync=true )
```
  public Function<Message<?>, Message<String>> borderflow_region2() {
        return (m) ->  {
            Charset cs = Charset.defaultCharset();
            System.out.println("charset="+cs);
            StringBuffer message = new StringBuffer();
            for(int i=0;i<540000;i++){
                message.append(new String("1".getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8));
            }
            String str = message.toString();
            System.out.println("length="+str.getBytes(StandardCharsets.UTF_8).length);
            return MessageBuilder.withPayload(str).build(); 
       };
    }

charset=UTF-8
length=540000
2021-11-04 15:24:26.720  WARN 1 --- [z2-borderflow-2] c.a.m.e.EventHubProducerAsyncClient      : Cannot send an EventBatch that is empty.
```

Change 54000 to 64000
Get the error of exceeding limit and warning of empty EventBatch
```
  public Function<Message<?>, Message<String>> borderflow_region2() {
        return (m) ->  {
            Charset cs = Charset.defaultCharset();
            System.out.println("charset="+cs);
            StringBuffer message = new StringBuffer();
            for(int i=0;i<640000;i++){
                message.append(new String("1".getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8));
            }
            String str = message.toString();
            System.out.println("length="+str.getBytes(StandardCharsets.UTF_8).length);
            return MessageBuilder.withPayload(str).build(); 
       };
  }


Change to 440000, no error nor waning:
```
  public Function<Message<?>, Message<String>> borderflow_region2() {
        return (m) ->  {
            Charset cs = Charset.defaultCharset();
            System.out.println("charset="+cs);
            StringBuffer message = new StringBuffer();
            for(int i=0;i<440000;i++){
                message.append(new String("1".getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8));
            }
            String str = message.toString();
            System.out.println("length="+str.getBytes(StandardCharsets.UTF_8).length);
            return MessageBuilder.withPayload(str).build(); 
       };
  }

charset=UTF-8
length=440000
````


charset=UTF-8
length=640000
2021-11-04 17:07:26.769  WARN 1 --- [z2-borderflow-2] c.a.messaging.eventhubs.EventDataBatch   : Size of the payload exceeded maximum message size: 1024 kb, errorContext[NAMESPACE: dev-qa-janus-events-cus.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: dev_qa_touchstone_borderflow_event, REFERENCE_ID: 4F17EA6F00F81B8800F81BBD6184134A_G10S3, LINK_CREDIT: 300]
2021-11-04 17:07:26.769 ERROR 1 --- [z2-borderflow-2] c.a.s.i.e.impl.AbstractEventHubTemplate  : Event is larger than maximum allowed size. Exception: com.azure.core.amqp.exception.AmqpException: Size of the payload exceeded maximum message size: 1024 kb, errorContext[NAMESPACE: dev-qa-janus-events-cus.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: dev_qa_touchstone_borderflow_event, REFERENCE_ID: 4F17EA6F00F81B8800F81BBD6184134A_G10S3, LINK_CREDIT: 300]
2021-11-04 17:07:26.769  WARN 1 --- [z2-borderflow-2] c.a.m.e.EventHubProducerAsyncClient      : Cannot send an EventBatch that is empty.
```


Use EventHubProducerClient to send 640000, no error nor warning:
```
public Function<Message<?>, Message<String>> borderflow_region2() {
        return (m) ->  {
            Charset cs = Charset.defaultCharset();
            System.out.println("charset="+cs);
            StringBuffer message = new StringBuffer();
            for(int i=0;i<640000;i++){
                message.append(new String("1".getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8));
            }
            String str = message.toString();
            System.out.println("length="+str.getBytes(StandardCharsets.UTF_8).length);

            EventData eventData = new EventData(str);
            EventDataBatch batch = eventHubProducerClient.createBatch();
            boolean isAdded = batch.tryAdd(eventData);
            System.out.println("isAdded="+isAdded);
            eventHubProducerClient.send(batch);

            return MessageBuilder.withPayload("1").build();
        };
    }


charset=UTF-8
length=640000
isAdded=true

```

**Comment by jialigit** [comment]
@w-lin hi, a little problem during debugging....The thing is this: I have tried to debug the source code, did not find any thing special that can change the EventData size. 

I want to know what is the difference between your project and my demo..
So, I hope you can give a short but complete demo that can reproduce this issue?   a demo repo is good.  

By the way,  what kind of subscription tier do you use? I hope to align with your project setting.

**Comment by jialigit** [comment]
Blocked for without user responding.

**Comment by backwind1233** [comment]
Closing this issue for long time no response.
